grakn-graph/src/main/java/ai/grakn/graph/internal/ElementFactory.java

```diff
@@ -185,7 +203,46 @@ RoleImpl buildRole(VertexElement vertex, Role type, Boolean isImplicit){
                     concept = new RuleTypeImpl(vertexElement);
                     break;
                 default:
-                    throw new RuntimeException("Unknown base type [" + vertexElement.label() + "]");
+                    throw GraphOperationException.unknownConcept(type.name());
+            }
+            graknGraph.txCache().cacheConcept(concept);
+        }
+        return graknGraph.txCache().getCachedConcept(conceptId);
+    }
+
+    /**
+     * Constructors are called directly because this is only called when reading a known {@link Edge} or {@link Concept}.
+     * Thus tracking the concept can be skipped.
+     *
+     * @param edge A {@link Edge} of an unknown type
+     * @return A concept built to the correct type
+     */
+    @Nullable
+    <X extends Concept> X buildConcept(Edge edge){
+        return buildConcept(buildEdgeElement(edge));
+    }
+
+    @Nullable
+    <X extends Concept> X buildConcept(EdgeElement edgeElement){
```

Comment:
This has a lot of duplicate code with it's brother `buildConcept(VertexElement)` need to clean that up in another PR